cmd/thanos/remote-write-receive.go

```diff
@@ -0,0 +1,281 @@
+	// initial config and mark ourselves as ready after it completed.
+	dbOpen := make(chan struct{})
+
+	// sync.Once is used to make sure we can close the channel at different execution stages(SIGTERM or when the config is loaded).
+	type closeOnce struct {
+		C     chan struct{}
+		once  sync.Once
+		Close func()
+	}
+	// Wait until the server is ready to handle reloading.
+	reloadReady := &closeOnce{
+		C: make(chan struct{}),
+	}
+	reloadReady.Close = func() {
+		reloadReady.once.Do(func() {
+			close(reloadReady.C)
+		})
+	}
+
+	level.Debug(logger).Log("msg", "setting up joining cluster")
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(func() error {
+			mint, maxt := metadata.Timestamps()
+			if peer != nil {
+				if err := peer.Join(cluster.PeerTypeSource, cluster.PeerMetadata{
+					Labels:  metadata.LabelsPB(),
+					MinTime: mint,
+					MaxTime: maxt,
+				}); err != nil {
+					return errors.Wrap(err, "join cluster")
+				}
+			}
+
+			<-ctx.Done()
+			return nil
+		}, func(err error) {
+			level.Debug(logger).Log("msg", "mesh group errored", "err", err)
+			cancel()
+			peer.Close(2 * time.Second)
+		})
+	}
+
+	level.Debug(logger).Log("msg", "setting up endpoint readiness")
+	{
+		// Initial configuration loading.
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				select {
+				case <-dbOpen:
+					break
+				// In case a shutdown is initiated before the dbOpen is released
+				case <-cancel:
+					reloadReady.Close()
+					return nil
+				}
+
+				reloadReady.Close()
+
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive web requests.")
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "initial load group errored", "err", err)
+				close(cancel)
+			},
+		)
+	}
+
+	level.Debug(logger).Log("msg", "setting up tsdb")
+	{
+		// TSDB.
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				level.Info(logger).Log("msg", "Starting TSDB ...")
+				db, err := tsdb.Open(
+					dataDir,
+					log.With(logger, "component", "tsdb"),
+					reg,
+					tsdbCfg,
+				)
+				if err != nil {
+					return fmt.Errorf("opening storage failed: %s", err)
+				}
+				level.Info(logger).Log("msg", "TSDB started")
+
+				startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+				localStorage.Set(db, startTimeMargin)
+				close(dbOpen)
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "tsdb group errored", "err", err)
+				if err := localStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(cancel)
+			},
+		)
+	}
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+	if err := metricHTTPListenGroup(g, logger, reg, httpMetricsBindAddr); err != nil {
+		level.Debug(logger).Log("msg", "metric listener errored", "err", err)
```

Comment:
Replicated error handling. (rule "Handle error just once") (: 

I think error wrap is the way to go (and removal of this log line).